edgedb-protocol/src/serialization/decode.rs

```diff
@@ -0,0 +1,14 @@
+mod raw_scalar;
+mod raw_composite;
+
+use snafu::ensure;
+use crate::errors::{self, DecodeError};
+
+pub use self::raw_composite::DecodeTupleLike;
+pub(crate) use self::raw_scalar::RawCodec;
+pub(crate) use self::raw_composite::{DecodeArrayLike, DecodeInputTuple};
+
+pub(crate) fn required_element(buf:Option<&[u8]>) -> Result<&[u8], DecodeError> {

```

Comment:
This is a weird function. It doesn't work on "elements" but rather on byte slices. 

Given that it's just onliner like `map(..).unwrap_or_else(...)` or `if let Some(...) = ...`, I would drop it. That would make error thrown more explicit.

Reply:
1. The `Option<&[u8]>` represents an element of a Shape or Tuple and the error code is specific to that interpretation.
2. I had it inline originally, but the nested mappings and error returns got pretty ugly inside `decode_tuple`

Reply:
`unwrap_or_else` expects an error, while `snafu` produces a result. So it's related to the same snafu problem mentioned in the other comment.